### ISSUE TITLE
simple custom metric from scikit-learn scorer

### ISSUE THREAD

**Issue by geoHeil** [Microsoft/LightGBM#844]
Scikit-learn provides several methods to calculate metrics from `(y_true, y_prediction)` i.e. f1, beta score like http://scikit-learn.org/stable/modules/generated/sklearn.metrics.fbeta_score.html https://github.com/Microsoft/LightGBM/issues/284 outlines how to build grad / hess objective, but is there a simpler way to simply use the output of i.e. the f_beta_score as the custom objective?

**Comment by Laurae2** [comment]
@geoHeil There's no simpler way as it requires to know the underlying mathematics.

To use it as a custom objective, you need the gradient (derivative) and hessian (second derivative) of your metric. Both of the derivative and second derivative must be convex (preferably), continuous (preferably), finite (preferably), and non-zero (zero will freeze the possible improvements as it is supposed to mean a perfect prediction).

Note that the gradient and hessian should be computed for each observation, and not globally. F-beta score would require to transform discrete gradient/hessian into continuous gradient/hessian first.

**Comment by geoHeil** [comment]
@Laurae2, thanks. The first part is clear to mea. Could you explain what
you mean with: Note that the gradient and hessian should be computed for
each observation, and not globally. F-beta score would require to transform
discrete gradient/hessian into continuous gradient/hessian first.

Laurae <notifications@github.com> schrieb am Fr., 18. Aug. 2017 um
11:42 Uhr:

**Comment by Laurae2** [comment]
@geoHeil 

> the gradient and hessian should be computed for each observation, and not globally

One observation = one pair of gradient/hessian

> F-beta score would require to transform
discrete gradient/hessian into continuous gradient/hessian first.

F-beta for a single observation is discrete, not continuous (it can take only a fixed amount of values => discrete).

In addition, as one observation = one pair of gradient/hessian, to transform F-beta to continuous, you must use the probabilities and not only the predicted class (which means you must also transform the metric).

**Comment by wxchan** [comment]
Maybe you can try something like https://github.com/HIPS/autograd, but I heard it's slow.

**Comment by geoHeil** [comment]
thanks
